### ISSUE TITLE
Could `copy_in` return an `impl Write`, so I can “push” data to the PostgreSQL server?

### ISSUE THREAD

**Issue by rory** [sfackler/rust-postgres#489]
The current [`copy_in`](https://docs.rs/postgres/0.16.0-rc.2/postgres/struct.Client.html#method.copy_in) and the the current released (v0.15.2) [`copy_in`](https://docs.rs/postgres/0.15.2/postgres/stmt/struct.Statement.html#method.copy_in) (which are for supporting PostgreSQL's `COPY ... FROM STDIN`) take a reader (i.e. something that `impl Read`). This allows one to e.g. read data from a file. the `rust-postgres` crate basically “pulls” data from the reader and send it along to the PostgreSQL server.

But in my use case, I am generating data that I want put into a postgres database, and this “pull”/read based approach is awkward. I would like if there was a copy function which returned a writer (e.g. `impl Write`), which I could write lines to (with e.g. `write!(...)`/`.write(...)`/`.write_all(...)`), allowing me to ”push” data to the PostgreSQL server, one line at a time. Currently I have a thing which `impl Read` by having a local buffer and pulling data from an `Iterator`, and there's a lot of data being copied around, as well as being complicated and challeging at first.. If I could do something like this, it would be much nicer:

    let mut copy_input_stream = connection.copy_in_writer(...)?;
    for line in my_data_generator() {
        write!(copy_input_stream, "{}\n", line)?;
    }
    copy_input_stream.finish()?;

If I write a programme which printed to stdout, and piped that to `psql` I would have this sort of ”push” functionality (I tried that, and it's an order of magnitude slower).

Would it be possible to make something like this? I haven't looked at the code, so I don't know if it's possible (nor looked at the v0.16+ code). I haven't used `copy_out`, but I presume the same suggestion could be raised there.

**Comment by sfackler** [comment]
I've been thinking about the same thing. Alternatively, instead of returning a writer, it could take in something like

```rust
trait WriteTo {
    fn write_to(self, w: &mut Write) -> io::Result<()>;
}
```

**Comment by rory** [comment]
I have made an attempt at implementing this [on my fork](https://github.com/rory/rust-postgres/tree/copy-in-writer). [here's an attempt at a diff](https://github.com/sfackler/rust-postgres/compare/master...rory:copy-in-writer).

I'm curious what feedback you have on this approach.

Annoyingly this method wasn't any faster for my programme! I had to wrap the returned Writer in a `BufWriter` which sped it up, looking at the reader code, it looks like that does that anyway. It does make the code cleaner and the library easier to use. I might try to clean this up for a proper pull request, but I forked v0.15.2 (the released version), and there is 0.16 which isn't released, and 0.17 which is under development, so I'm unsure what I should do.
